pkg/types/builder.go

```diff
@@ -282,23 +314,51 @@ func (r *resource) addParameterField(f *Field, field *types.Var) {
 	// - req => required
 	// - !f.Identifier => not identifiers - i.e. region, zone, etc.
 	// - len(f.CanonicalPaths) == 1 => top level, i.e. not a nested field
+	// TODO (lsviben): We should add CEL rules for all required fields,
+	// not just the top level ones, due to having all forProvider
+	// fields now optional. CEL rules should check if a field is
+	// present either in forProvider or initProvider.
 	if req && !f.Identifier && len(f.CanonicalPaths) == 1 {
 		req = false
-		r.topLevelRequiredParams = append(r.topLevelRequiredParams, f.TransformedName)
+		// If the field is not a terraform field, we should not require it in init,
+		// as it is not an initProvider field.
+		r.topLevelRequiredParams = append(r.topLevelRequiredParams, newTopLevelRequiredParam(f.TransformedName, f.TFTag != "-"))
 	}
 
-	f.Comment.Required = &req
-	if !req {
-		r.paramTags = append(r.paramTags, fmt.Sprintf(`json:"%s" tf:"%s"`, f.JSONTag, f.TFTag))
-	} else {
+	// Note(lsviben): Only identifier fields should be required,
+	// as they are unique for spec.forProvider.
+	if req && f.Identifier {
 		// Required fields should not have omitempty tag in json tag.
 		// TODO(muvaf): This overrides user intent if they provided custom
 		// JSON tag.
 		r.paramTags = append(r.paramTags, fmt.Sprintf(`json:"%s" tf:"%s"`, strings.TrimSuffix(f.JSONTag, ",omitempty"), f.TFTag))
+	} else {
+		r.paramTags = append(r.paramTags, fmt.Sprintf(`json:"%s" tf:"%s"`, f.JSONTag, f.TFTag))
```

Comment:
Not sure exactly where but looks like we lost `+kubebuilder:validation:Required` for identifier arguments: https://github.com/upbound/provider-aws/pull/785/files#diff-7a8cb7fd3694d62f4b9b059131d2d57c12546cab7142c46c26c1d08ee87dc7e2L40

<img width="530" alt="Screenshot 2023-07-24 at 14 43 42" src="https://github.com/upbound/upjet/assets/9900707/137012ba-ce5f-4771-9b09-4444192d3fb8">

Reply:
Seems i removed setting the required comment. Reverted it now and modified to be required if its required and the field is not an initProvider field

Reply:
Ok, now I am seeing the required marker is back for identifier but optional ones are still gone. Any reason not to keep `// +kubebuilder:validation:Optional` markers for optional parameters?

Reply:
Returned them. 

We could also put the `// +kubebuilder:validation:Optional` on `initProvider` and `atProvider` fields if we want. `atProvider` did not have them earlier though. Not sure what difference it makes, it seems they are optional with or without them